Sparsemax

AndreasMadsen opened the pull request:
The sparsemax op is an alternative to the softmax op, that allows the
output to be sparse (zero properbility) while stil sharing many
mathematical properties with softmax.

The cross entropy loss doesn't work with sparsemax as log(0) is not
defined, thus there is also a sparsemax loss function. This loss
function have a gradient equivalent to that of cross entropy when
using softmax.

Original sparsemax article: https://arxiv.org/abs/1602.02068

----

I had some issues with getting the numerical precision good enough for `assertAllCloseAccordingToType`. subtracting `mean(logits)` helped a bit but not enough, thus I have expanded the assert method such that the tolerance can be specified.

----

_This code was developed by me (@andreasmadsen), @FrederikWR and @MarcoDalFarra. The original code can be found in https://github.com/AndreasMadsen/course-02456-sparsemax. The project was supervised by @alrojo._

AndreasMadsen committed:
allow specifying tol for each dtype in equal test

AndreasMadsen committed:
implement sparsemax and sparsemax loss ops

The sparsemax op is an alternative to the softmax op, that allows the
output to be sparse (zero probability) while stil sharing many
mathematical properties with softmax.

The cross entropy loss doesn't work with sparsemax as log(0) is not
defined, thus there is also a sparsemax loss function. This loss
function have a gradient equivalent to that of cross entropy when
using softmax.

Original sparsemax article: https://arxiv.org/abs/1602.02068

tensorflow-jenkins commented:
Can one of the admins verify this patch?

jhseu commented:
Thanks for the pull request! It's very well written.

For now, we think this should live outside of the official repo. We can add it later if it proves to be useful to the wider community.

AndreasMadsen commented:
We where originally informed by @alrojo, who had talked to some TensorFlow members, that this was of interrest. It must have been a misunderstanding.

jhseu commented:
Ah, do you mind mentioning who you talked to? Perhaps they can review.

tensorflow-jenkins commented:
Can one of the admins verify this patch?

lukaszkaiser commented:
Hi guys. I talked to alrojo and I'm happy to try to move this forward.

One main thing that came up is how we could do it without the extra kernels (which is a large burden to support). Here is a question to you guys:

It seems like a lot of the work is because you need sorting. But tf.nn.top_k can be used to sort. Did you consider using it? Could you try it to simplify the code?

AndreasMadsen commented:
> It seems like a lot of the work is because you need sorting. But tf.nn.top_k can be used to sort. Did you consider using it? Could you try it to simplify the code?

I think this is true, there are two issues as far as I'm aware:

* `tf.nn.top_k` has no GPU implementation.
* I don't know how to define a gradient on a composite operator (not a kernel, but an graph).

On the subject of simplifying code: I don't think subtracting the mean is particular important, I helps only very little in some situations. Unlike softmax, sparsemax mostly consists of additions and some multiplications (no `exp`). So this part could likely be removed without noticeable issues. In fact we didn't include it in our experiments.

lukaszkaiser commented:
I don't think we need to worry about top_k not having a GPU kernel-- it's a core op, so adding kernels to it is more straightforward than adding (and supporting and maintaining) a few new kernels in contrib.

As for gradients, it can be done with Defun and grad_func, see for example https://github.com/tensorflow/tensorflow/issues/4661. But I'm curious why it is needed at all?

AndreasMadsen committed:
reimplement sparsemax ops in just python

AndreasMadsen commented:
> I don't think we need to worry about top_k not having a GPU kernel-- it's a core op, so adding kernels to it is more straightforward than adding (and supporting and maintaining) a few new kernels in contrib.

Originally we implemented sparsemax from a performance perspective, not from a maintenance perspective.

I've replaced the kernels such the implementations just use core ops (see https://github.com/tensorflow/tensorflow/pull/6387/commits/ccf12a892d4a9b4f3166e5818374bf65e7a13195). One issue I couldn't solve was how to extract the cumsum values without depending on the batch size `.get_shape()[0]` (see https://github.com/tensorflow/tensorflow/pull/6387/commits/ccf12a892d4a9b4f3166e5818374bf65e7a13195#diff-e786beafb4ecbf8faeae13d5d3d75ea4R66).

> As for gradients, it can be done with Defun and grad_func, see for example #4661. But I'm curious why it is needed at all?

It is not necessary as such, but the gradients of `sparsemax` and `sparsemax_loss` are very simple. I assume it will be more performant to express them directly, than just using backprop though a fairly complex graph. I tried to implement them using `Defun` as you sugested, unfortunately I get `ValueError: Cannot convert an unknown Dimension to a Tensor: ?` errors. Can you take a look at https://github.com/tensorflow/tensorflow/pull/6387/commits/ca605df0991d4de2f4f78c7381e9ed30d53548cf ?


<details>
  <summary>full error message</summary>

```
$ PYTHONPATH=tensorflow/_python_build python3 tensorflow/tensorflow/contrib/sparsemax/python/kernel_tests/sparsemax_test.py

/Users/Andreas/Sites/tensorflow/_python_build/tensorflow/python/framework/function.py:969: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() instead
  argspec = inspect.getargspec(func)
EE.
======================================================================
ERROR: testDouble (__main__.SparsemaxTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/Andreas/Sites/tensorflow/_python_build/tensorflow/python/framework/op_def_library.py", line 491, in apply_op
    preferred_dtype=default_dtype)
  File "/Users/Andreas/Sites/tensorflow/_python_build/tensorflow/python/framework/ops.py", line 716, in internal_convert_to_tensor
    ret = conversion_func(value, dtype=dtype, name=name, as_ref=as_ref)
  File "/Users/Andreas/Sites/tensorflow/_python_build/tensorflow/python/framework/constant_op.py", line 212, in _dimension_tensor_conversion_function
    raise ValueError("Cannot convert an unknown Dimension to a Tensor: %s" % d)
ValueError: Cannot convert an unknown Dimension to a Tensor: ?

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "tensorflow/tensorflow/contrib/sparsemax/python/kernel_tests/sparsemax_test.py", line 258, in testDouble
    self._test_dtype('float64')
  File "tensorflow/tensorflow/contrib/sparsemax/python/kernel_tests/sparsemax_test.py", line 224, in _test_dtype
    self._test_sparsemax_against_numpy(dtype, random, use_gpu=False)
  File "tensorflow/tensorflow/contrib/sparsemax/python/kernel_tests/sparsemax_test.py", line 83, in _test_sparsemax_against_numpy
    tf_sparsemax_op, tf_sparsemax_out = self._tf_sparsemax(z, dtype, use_gpu)
  File "tensorflow/tensorflow/contrib/sparsemax/python/kernel_tests/sparsemax_test.py", line 74, in _tf_sparsemax
    tf_sparsemax_op = sparsemax(z.astype(dtype))
  File "/Users/Andreas/Sites/tensorflow/_python_build/tensorflow/python/framework/function.py", line 885, in __call__
    return self.instantiate(input_types)(*args, **kwargs)
  File "/Users/Andreas/Sites/tensorflow/_python_build/tensorflow/python/framework/function.py", line 860, in instantiate
    _ = defined.name  # Fully instantiate the function definition.
  File "/Users/Andreas/Sites/tensorflow/_python_build/tensorflow/python/framework/function.py", line 608, in name
    self._create_definition_if_needed()
  File "/Users/Andreas/Sites/tensorflow/_python_build/tensorflow/python/framework/function.py", line 659, in _create_definition_if_needed
    outputs = self._func(*inputs)
  File "/Users/Andreas/Sites/tensorflow/_python_build/tensorflow/contrib/sparsemax/python/ops/sparsemax.py", line 90, in sparsemax
    return fprop(logits)
  File "/Users/Andreas/Sites/tensorflow/_python_build/tensorflow/python/framework/function.py", line 885, in __call__
    return self.instantiate(input_types)(*args, **kwargs)
  File "/Users/Andreas/Sites/tensorflow/_python_build/tensorflow/python/framework/function.py", line 860, in instantiate
    _ = defined.name  # Fully instantiate the function definition.
  File "/Users/Andreas/Sites/tensorflow/_python_build/tensorflow/python/framework/function.py", line 608, in name
    self._create_definition_if_needed()
  File "/Users/Andreas/Sites/tensorflow/_python_build/tensorflow/python/framework/function.py", line 659, in _create_definition_if_needed
    outputs = self._func(*inputs)
  File "/Users/Andreas/Sites/tensorflow/_python_build/tensorflow/contrib/sparsemax/python/ops/sparsemax.py", line 67, in fprop
    z_sorted, _ = nn.top_k(z, k=dims)
  File "/Users/Andreas/Sites/tensorflow/_python_build/tensorflow/python/ops/nn_ops.py", line 1946, in top_k
    return gen_nn_ops._top_kv2(input, k=k, sorted=sorted, name=name)
  File "/Users/Andreas/Sites/tensorflow/_python_build/tensorflow/python/ops/gen_nn_ops.py", line 2464, in _top_kv2
    name=name)
  File "/Users/Andreas/Sites/tensorflow/_python_build/tensorflow/python/framework/op_def_library.py", line 504, in apply_op
    values, as_ref=input_arg.is_ref).dtype.name
  File "/Users/Andreas/Sites/tensorflow/_python_build/tensorflow/python/framework/ops.py", line 716, in internal_convert_to_tensor
    ret = conversion_func(value, dtype=dtype, name=name, as_ref=as_ref)
  File "/Users/Andreas/Sites/tensorflow/_python_build/tensorflow/python/framework/constant_op.py", line 212, in _dimension_tensor_conversion_function
    raise ValueError("Cannot convert an unknown Dimension to a Tensor: %s" % d)
ValueError: Cannot convert an unknown Dimension to a Tensor: ?

======================================================================
ERROR: testFloat (__main__.SparsemaxTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/Andreas/Sites/tensorflow/_python_build/tensorflow/python/framework/op_def_library.py", line 491, in apply_op
    preferred_dtype=default_dtype)
  File "/Users/Andreas/Sites/tensorflow/_python_build/tensorflow/python/framework/ops.py", line 716, in internal_convert_to_tensor
    ret = conversion_func(value, dtype=dtype, name=name, as_ref=as_ref)
  File "/Users/Andreas/Sites/tensorflow/_python_build/tensorflow/python/framework/constant_op.py", line 212, in _dimension_tensor_conversion_function
    raise ValueError("Cannot convert an unknown Dimension to a Tensor: %s" % d)
ValueError: Cannot convert an unknown Dimension to a Tensor: ?

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "tensorflow/tensorflow/contrib/sparsemax/python/kernel_tests/sparsemax_test.py", line 255, in testFloat
    self._test_dtype('float32')
  File "tensorflow/tensorflow/contrib/sparsemax/python/kernel_tests/sparsemax_test.py", line 224, in _test_dtype
    self._test_sparsemax_against_numpy(dtype, random, use_gpu=False)
  File "tensorflow/tensorflow/contrib/sparsemax/python/kernel_tests/sparsemax_test.py", line 83, in _test_sparsemax_against_numpy
    tf_sparsemax_op, tf_sparsemax_out = self._tf_sparsemax(z, dtype, use_gpu)
  File "tensorflow/tensorflow/contrib/sparsemax/python/kernel_tests/sparsemax_test.py", line 74, in _tf_sparsemax
    tf_sparsemax_op = sparsemax(z.astype(dtype))
  File "/Users/Andreas/Sites/tensorflow/_python_build/tensorflow/python/framework/function.py", line 885, in __call__
    return self.instantiate(input_types)(*args, **kwargs)
  File "/Users/Andreas/Sites/tensorflow/_python_build/tensorflow/python/framework/function.py", line 860, in instantiate
    _ = defined.name  # Fully instantiate the function definition.
  File "/Users/Andreas/Sites/tensorflow/_python_build/tensorflow/python/framework/function.py", line 608, in name
    self._create_definition_if_needed()
  File "/Users/Andreas/Sites/tensorflow/_python_build/tensorflow/python/framework/function.py", line 659, in _create_definition_if_needed
    outputs = self._func(*inputs)
  File "/Users/Andreas/Sites/tensorflow/_python_build/tensorflow/contrib/sparsemax/python/ops/sparsemax.py", line 90, in sparsemax
    return fprop(logits)
  File "/Users/Andreas/Sites/tensorflow/_python_build/tensorflow/python/framework/function.py", line 885, in __call__
    return self.instantiate(input_types)(*args, **kwargs)
  File "/Users/Andreas/Sites/tensorflow/_python_build/tensorflow/python/framework/function.py", line 860, in instantiate
    _ = defined.name  # Fully instantiate the function definition.
  File "/Users/Andreas/Sites/tensorflow/_python_build/tensorflow/python/framework/function.py", line 608, in name
    self._create_definition_if_needed()
  File "/Users/Andreas/Sites/tensorflow/_python_build/tensorflow/python/framework/function.py", line 659, in _create_definition_if_needed
    outputs = self._func(*inputs)
  File "/Users/Andreas/Sites/tensorflow/_python_build/tensorflow/contrib/sparsemax/python/ops/sparsemax.py", line 67, in fprop
    z_sorted, _ = nn.top_k(z, k=dims)
  File "/Users/Andreas/Sites/tensorflow/_python_build/tensorflow/python/ops/nn_ops.py", line 1946, in top_k
    return gen_nn_ops._top_kv2(input, k=k, sorted=sorted, name=name)
  File "/Users/Andreas/Sites/tensorflow/_python_build/tensorflow/python/ops/gen_nn_ops.py", line 2464, in _top_kv2
    name=name)
  File "/Users/Andreas/Sites/tensorflow/_python_build/tensorflow/python/framework/op_def_library.py", line 504, in apply_op
    values, as_ref=input_arg.is_ref).dtype.name
  File "/Users/Andreas/Sites/tensorflow/_python_build/tensorflow/python/framework/ops.py", line 716, in internal_convert_to_tensor
    ret = conversion_func(value, dtype=dtype, name=name, as_ref=as_ref)
  File "/Users/Andreas/Sites/tensorflow/_python_build/tensorflow/python/framework/constant_op.py", line 212, in _dimension_tensor_conversion_function
    raise ValueError("Cannot convert an unknown Dimension to a Tensor: %s" % d)
ValueError: Cannot convert an unknown Dimension to a Tensor: ?

----------------------------------------------------------------------
Ran 3 tests in 0.029s

FAILED (errors=2)
```
</details>

lukaszkaiser commented:
Great thanks for the new implementation, looks much more maintainable! Just a few more things.

* You wrote: " how to extract the cumsum values without depending on the batch size" -- I'm not sure I understand. Do you need the batch size to be a python integer rather than a tensor? I mean, instead of `.get_shape()` why not use `tf.shape`, would that work?

* I understand that you implemented it with performance in mind first, and I think that's great. Could you compare the performance of your hand-written kernels and the new implementation that only uses core ops? I'm very curious how much is getting lost here -- if it's 20% then maybe it's ok, but if it's more then we might need to think how to speed things up.

* As for your error, it looks like `dims = logits.get_shape()[1]` is `None` and this makes things fail. Again -- you're trying to get the shape statically, and that's not always available. Could you try `tf.shape` instead?

* Finally, if you have time and after everything works, I'd be very curious how much faster the hand-written backprop is compared to TF-computed backprop. It might be an interesting case for the new TF compiler.

Great thanks again for doing all of this!

AndreasMadsen commented:
> You wrote: " how to extract the cumsum values without depending on the batch size" -- I'm not sure I understand. Do you need the batch size to be a python integer rather than a tensor? I mean, instead of .get_shape() why not use tf.shape, would that work?

Thanks `tf.shape` works, I was just not aware of it.

> I understand that you implemented it with performance in mind first, and I think that's great. Could you compare the performance of your hand-written kernels and the new implementation that only uses core ops? I'm very curious how much is getting lost here -- if it's 20% then maybe it's ok, but if it's more then we might need to think how to speed things up.

> Finally, if you have time and after everything works, I'd be very curious how much faster the hand-written backprop is compared to TF-computed backprop. It might be an interesting case for the new TF compiler.

Yes, once I get the hand-written backprop version to work I will benchmark the diffrent versions.

> As for your error, it looks like dims = logits.get_shape()[1] is None and this makes things fail. Again -- you're trying to get the shape statically, and that's not always available. Could you try tf.shape instead?

I changed it to  `tf.shape`, now I get a strange recursion error: [full error message](https://gist.github.com/AndreasMadsen/8dca7e8b3ead6b5b339fd4a8c5bc8b03).

lukaszkaiser commented:
Oh, I think you can't use fprop in bprop when fprop is decorated with bprop like you have. Just make an fprop_raw function, same as your fprop but without @Defun, and call that one in bprop and fprop.

Hope things will work soon, looking forward to it!

AndreasMadsen commented:
> Oh, I think you can't use fprop in bprop when fprop is decorated with bprop like you have. Just make an fprop_raw function, same as your fprop but without `@Defun`, and call that one in bprop and fprop.

That was it. It a little strange, as #4661 is about depending on `fprop` in `bprop`.

Anyway, I'm getting a new error.

```
PYTHONPATH=tensorflow/_python_build python3 tensorflow/tensorflow/contrib/sparsemax/python/kernel_tests/sparsemax_test.py
/Users/Andreas/Sites/tensorflow/_python_build/tensorflow/python/framework/function.py:969: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() instead
  argspec = inspect.getargspec(func)
EE.
======================================================================
ERROR: testDouble (__main__.SparsemaxTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tensorflow/tensorflow/contrib/sparsemax/python/kernel_tests/sparsemax_test.py", line 258, in testDouble
    self._test_dtype('float64')
  File "tensorflow/tensorflow/contrib/sparsemax/python/kernel_tests/sparsemax_test.py", line 224, in _test_dtype
    self._test_sparsemax_against_numpy(dtype, random, use_gpu=False)
  File "tensorflow/tensorflow/contrib/sparsemax/python/kernel_tests/sparsemax_test.py", line 88, in _test_sparsemax_against_numpy
    self.assertShapeEqual(p_sparemax, tf_sparsemax_op)
  File "/Users/Andreas/Sites/tensorflow/_python_build/tensorflow/python/framework/test_util.py", line 594, in assertShapeEqual
    self.assertAllEqual(np_array.shape, tf_tensor.get_shape().as_list())
  File "/Users/Andreas/Sites/tensorflow/_python_build/tensorflow/python/framework/tensor_shape.py", line 782, in as_list
    raise ValueError("as_list() is not defined on an unknown TensorShape.")
ValueError: as_list() is not defined on an unknown TensorShape.
```

I don't understand why the TensorShape would be unknown, as the sparsemax op is passed the numpy array directly. The tests also worked for both the kernel and `name_scope` implementations.

AndreasMadsen commented:
Here is an initial benchmark:

| size & case                                | kernel CPU     | name_scope CPU   |
|:--------------------------------|:---------------|:-----------------|
| (1000, 10) sparsemax            | 0.052 ± 0.0032 | 0.203 ± 0.0364   |
| (1000, 10) sparsemax_grad       | 0.071 ± 0.0009 | 0.300 ± 0.0122   |
| (1000, 10) sparsemax_loss       | 0.060 ± 0.0021 | 0.213 ± 0.0112   |
| (1000, 10) sparsemax_loss_grad  | 0.066 ± 0.0015 | 0.310 ± 0.0127   |
| (1000, 100) sparsemax           | 0.293 ± 0.0094 | 0.988 ± 0.0689   |
| (1000, 100) sparsemax_grad      | 0.373 ± 0.0172 | 1.410 ± 0.0110   |
| (1000, 100) sparsemax_loss      | 0.330 ± 0.0085 | 0.979 ± 0.0119   |
| (1000, 100) sparsemax_loss_grad | 0.447 ± 0.0158 | 1.597 ± 0.0159   |

_`size` specifies the size of the logits and label matrix, both where randomly generated. For each case, 100 iterations where performed to avoid clock issues and `tf.Variable().initializer` was used to avoid a memory transfer overhead. Finally each experiment was repeated 10 times, the summary of those repetitions are shown as `mean ± 95% confidence interval`._

lukaszkaiser commented:
Ok, so can I assume that the pure-TF-ops implementation works now? That's good!

The benchmarks look a bit disappointing. But I see 2 things to try. For one, did you take a look at the timeline? Maybe it'll make it more clear what's eating the time. The other thing is that we should probably focus on larger sizes, how about (16K, 512) or so?

Thanks again for doing this! I think the pure-TF code is basically ready for review and checking-in, but it might be useful to think a bit more about performance first.

AndreasMadsen commented:
> Ok, so can I assume that the pure-TF-ops implementation works now? That's good!

The Defun version doesn't work. I still get the error from https://github.com/tensorflow/tensorflow/pull/6387#issuecomment-269468829.

lukaszkaiser commented:
Hmm, are you sure you want `self.assertShapeEqual(p_sparemax, tf_sparsemax_op)` and not `self.assertShapeEqual(p_sparemax, tf_sparsemax_out)` there? I though you want to compare the numpy shapes? I'm very curious if the grad_func helps solve some performance problems!

lukaszkaiser commented:
You could do a `tf.set_shape` at the end to set the shape of the Tensor properly. Or just use assertEqual on x.shape and y.shape if x and y are numpy objects.

AndreasMadsen commented:
> You could do a tf.set_shape at the end to set the shape of the Tensor properly. Or just use assertEqual on x.shape and y.shape if x and y are numpy objects.

I find it really odd that I have to do that, but okay.

Here are the benchmark results:

|                                 | kernel CPU     | kernel GPU      | name_scope CPU   | Defun CPU       |
|:--------------------------------|:---------------|:----------------|:-----------------|:----------------|
| (1000, 10) sparsemax            | 0.097 ± 0.0053 | 0.110 ± 0.0645  | 0.347 ± 0.0573   | 0.372 ± 0.0941  |
| (1000, 10) sparsemax_grad       | 0.163 ± 0.0022 | 0.109 ± 0.0032  | 0.534 ± 0.0066   | 0.687 ± 0.0093  |
| (1000, 10) sparsemax_loss       | 0.133 ± 0.0037 | 0.105 ± 0.0029  | 0.358 ± 0.0047   | 0.399 ± 0.0043  |
| (1000, 10) sparsemax_loss_grad  | 0.132 ± 0.0033 | 0.091 ± 0.0014  | 0.573 ± 0.0078   | 0.959 ± 0.0118  |
| (1000, 100) sparsemax           | 0.602 ± 0.0535 | 0.221 ± 0.0042  | 1.509 ± 0.0626   | 1.494 ± 0.0701  |
| (1000, 100) sparsemax_grad      | 1.045 ± 0.0275 | 0.238 ± 0.0046  | 2.429 ± 0.0511   | 2.967 ± 0.0461  |
| (1000, 100) sparsemax_loss      | 0.819 ± 0.0410 | 0.281 ± 0.0044  | 1.741 ± 0.0390   | 1.739 ± 0.0420  |
| (1000, 100) sparsemax_loss_grad | 0.984 ± 0.0472 | 0.227 ± 0.0031  | 2.634 ± 0.0455   | 3.752 ± 0.0364  |
| (1000, 512) sparsemax           | 3.180 ± 0.0149 | 1.739 ± 0.0463  | 6.008 ± 0.0546   | 5.677 ± 0.0560  |
| (1000, 512) sparsemax_grad      | 3.972 ± 0.0260 | 1.779 ± 0.0169  | 8.192 ± 0.0301   | 11.660 ± 0.0233 |
| (1000, 512) sparsemax_loss      | 3.496 ± 0.0271 | 1.688 ± 0.0088  | 6.319 ± 0.0404   | 6.370 ± 0.0203  |
| (1000, 512) sparsemax_loss_grad | 3.807 ± 0.0859 | 1.720 ± 0.0073  | 9.001 ± 0.0540   | 13.823 ± 0.0294 |
| (20, 16384) sparsemax           | 2.748 ± 0.0427 | 26.740 ± 0.0171 | 4.808 ± 0.0598   | 4.691 ± 0.0679  |
| (20, 16384) sparsemax_grad      | 3.417 ± 0.0588 | 26.715 ± 0.0147 | 6.217 ± 0.0557   | 9.567 ± 0.0327  |
| (20, 16384) sparsemax_loss      | 3.148 ± 0.0744 | 27.026 ± 0.0262 | 5.056 ± 0.0366   | 5.272 ± 0.0159  |
| (20, 16384) sparsemax_loss_grad | 3.422 ± 0.0411 | 26.772 ± 0.0108 | 6.851 ± 0.0401   | 11.347 ± 0.0505 |

lukaszkaiser commented:
Some strange things. Why do you run (20, 16K)? I thought 16K labels with dense tensors of size 512 is more realistic?

But, more importantly, it's strange that Defun makes things slow. When benchmarking TF one always has to first run the benchmark a few hundred steps without measuring, just to warm-up the graph and stuff. Are you doing that?

AndreasMadsen commented:
> Why do you run (20, 16K)? I thought 16K labels with dense tensors of size 512 is more realistic?

I just don't have the resources to run any bigger.

> When benchmarking TF one always has to first run the benchmark a few hundred steps without measuring, just to warm-up the graph and stuff. Are you doing that?

This is the benchmark suite I wrote https://github.com/AndreasMadsen/tensorflow-sparsemax-benchmark. Do you have an example of how to ideally benchmark a TF function, I think that would be more productive.

lukaszkaiser commented:
I think there is something I don't understand about how you specify sizes. A normal softmax from vectors of size 512 into 16K labels easily fits into any GPU, so I'm clearly not understanding something here.

But as for your benchmarks: you should always first run at least one iteration and throw away the timing, only later start true measurements. The graph does a fair bit of one-time allocations, especially with Defun, no need to measure them.

Give it another look, if the benchmarks still show that the Defun isn't helping, then I think we should check in the code without it and focus on optimizing that. Did you look at the TF timeline to see what takes so much time? (We can optimize performance after checking this in too, as you prefer.)

AndreasMadsen commented:
> But as for your benchmarks: you should always first run at least one iteration and throw away the timing, only later start true measurements. The graph does a fair bit of one-time allocations, especially with Defun, no need to measure them.

I've changed it such it runs it 100 times before it measures anything. It also doesn't measure the memory transfer anymore.

> I think there is something I don't understand about how you specify sizes. A normal softmax from vectors of size 512 into 16K labels easily fits into any GPU, so I'm clearly not understanding something here.

It was a matter of time, not memory. I created a job script that I can submit to our HPC system so that is not a problem, but it literally takes 12h to run the benchmark. 

|                                  | kernel CPU      | kernel GPU       | name_scope CPU   | Defun CPU        |
|:---------------------------------|:----------------|:-----------------|:-----------------|:-----------------|
| (1000, 10) sparsemax             | 0.096 ± 0.0008  | 0.079 ± 0.0009   | 0.325 ± 0.0032   | 0.336 ± 0.0030   |
| (1000, 10) sparsemax_grad        | 0.163 ± 0.0009  | 0.106 ± 0.0014   | 0.536 ± 0.0061   | 0.685 ± 0.0045   |
| (1000, 10) sparsemax_loss        | 0.131 ± 0.0020  | 0.104 ± 0.0015   | 0.356 ± 0.0039   | 0.399 ± 0.0039   |
| (1000, 10) sparsemax_loss_grad   | 0.129 ± 0.0031  | 0.088 ± 0.0005   | 0.573 ± 0.0040   | 0.950 ± 0.0050   |
| (1000, 100) sparsemax            | 0.554 ± 0.0571  | 0.213 ± 0.0013   | 1.494 ± 0.0316   | 1.493 ± 0.0291   |
| (1000, 100) sparsemax_grad       | 1.029 ± 0.0267  | 0.237 ± 0.0029   | 2.399 ± 0.0495   | 3.067 ± 0.0318   |
| (1000, 100) sparsemax_loss       | 0.844 ± 0.0182  | 0.277 ± 0.0034   | 1.758 ± 0.0616   | 1.735 ± 0.0159   |
| (1000, 100) sparsemax_loss_grad  | 0.994 ± 0.0278  | 0.227 ± 0.0009   | 2.632 ± 0.0245   | 3.723 ± 0.0316   |
| (1000, 512) sparsemax            | 3.238 ± 0.0351  | 1.746 ± 0.0010   | 5.973 ± 0.0187   | 5.618 ± 0.0239   |
| (1000, 512) sparsemax_grad       | 3.918 ± 0.0288  | 1.793 ± 0.0019   | 8.079 ± 0.0312   | 11.630 ± 0.0235  |
| (1000, 512) sparsemax_loss       | 3.480 ± 0.0181  | 1.673 ± 0.0016   | 6.218 ± 0.0433   | 6.309 ± 0.0219   |
| (1000, 512) sparsemax_loss_grad  | 3.887 ± 0.0627  | 1.754 ± 0.0048   | 8.893 ± 0.0575   | 13.813 ± 0.0495  |
| (512, 16384) sparsemax           | 60.092 ± 1.4167 | 633.671 ± 0.2451 | 86.176 ± 0.9363  | 91.747 ± 0.9430  |
| (512, 16384) sparsemax_grad      | 69.699 ± 0.9295 | 634.124 ± 0.3004 | 115.631 ± 1.1464 | 187.159 ± 1.0400 |
| (512, 16384) sparsemax_loss      | 61.021 ± 0.7513 | 632.289 ± 0.2675 | 94.075 ± 0.6780  | 97.617 ± 0.2923  |
| (512, 16384) sparsemax_loss_grad | 67.598 ± 1.7163 | 633.777 ± 0.3879 | 125.269 ± 0.8877 | 213.390 ± 1.1162 |

> Did you look at the TF timeline to see what takes so much time? 

No I have not, I want to setup the benchmarks correctly before I profile it.

lukaszkaiser commented:
Ok, thanks! It looks like the Defun isn't worth it. Can we remove it and check in the code? I'll ask about top_k GPU kernel in the meantime.

AndreasMadsen commented:
> Ok, thanks! It looks like the Defun isn't worth it. Can we remove it and check in the code? I'll ask about top_k GPU kernel in the meantime.

I've removed the defun commit.

lukaszkaiser commented on tensorflow/python/framework/test_util.py:
```diff
@@ -473,7 +473,9 @@ def assertAllClose(self, a, b, rtol=1e-6, atol=1e-6):
       np.testing.assert_allclose(a, b, rtol=rtol, atol=atol)
 
-  def assertAllCloseAccordingToType(self, a, b, rtol=1e-6, atol=1e-6):
+  def assertAllCloseAccordingToType(self, a, b, rtol=1e-6, atol=1e-6,
```
Please add a unit test for this function in /tensorflow/python/framework/tensor_util_test.py.

lukaszkaiser replied:
Sorry, I meant test_util_test of course.

lukaszkaiser reviewed (changes requested):
Looks good to me, but please add one more unit test for the changed test function.

lukaszkaiser commented:
Ok, thanks! I think the code is ready, but please add one more short unit test for the changed `assertAllCloseAccordingToType` function (just so it's at least ran once in the tests).

AndreasMadsen commented:
added tests for `assertAllCloseAccordingToType` as requested.

lukaszkaiser reviewed (approved):
Thanks, looks good.

scottclifford commented on tensorflow/contrib/sparsemax/python/kernel_tests/sparsemax_test.py:
```diff
@@ -0,0 +1,252 @@
+    grad = np.ones_like(z)
+
+    # Construct S(z)
+    properbility = self._np_sparsemax(z)
```
probability?

drpngx replied:
Also please take take a look at this

lukaszkaiser commented:
@tensorflow-jenkins test this please.

lukaszkaiser commented:
@tensorflow-jenkins test this please

lukaszkaiser commented:
Looks like lint checks are failing due to some bad indentation:

tensorflow/python/framework/test_util_test.py:198: [W0311(bad-indentation), ] Bad indentation. Found 6 spaces, expected 4

Could you correct that? Thanks!

AndreasMadsen commented:
Sure, what is the command for running just the lint check?

lukaszkaiser commented:
Take a look here (search for FAIL):

https://ci.tensorflow.org/job/tensorflow-pull-requests-sanity/2635/console

drpngx commented:
@AndreasMadsen any luck?

```
FAIL: Found 9 non-whitelited pylint errors:
tensorflow/python/framework/test_util_test.py:198: [W0311(bad-indentation), ] Bad indentation. Found 6 spaces, expected 4

tensorflow/python/framework/test_util_test.py:204: [W0311(bad-indentation), ] Bad indentation. Found 6 spaces, expected 4

tensorflow/python/framework/test_util_test.py:205: [W0311(bad-indentation), ] Bad indentation. Found 10 spaces, expected 6

tensorflow/python/framework/test_util_test.py:212: [W0311(bad-indentation), ] Bad indentation. Found 6 spaces, expected 4

tensorflow/python/framework/test_util_test.py:219: [W0311(bad-indentation), ] Bad indentation. Found 6 spaces, expected 4

tensorflow/python/framework/test_util_test.py:220: [W0311(bad-indentation), ] Bad indentation. Found 10 spaces, expected 6

tensorflow/python/framework/test_util_test.py:228: [W0311(bad-indentation), ] Bad indentation. Found 6 spaces, expected 4

tensorflow/python/framework/test_util_test.py:236: [W0311(bad-indentation), ] Bad indentation. Found 6 spaces, expected 4

tensorflow/python/framework/test_util_test.py:237: [W0311(bad-indentation), ] Bad indentation. Found 10 spaces, expected 6

```

AndreasMadsen commented:
@drpngx No. Finding the errors is no problem, but I would like to lint the files myself before committing, unfortunately I can't find the appropriate command.

drpngx commented:
Gotcha.
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/ci_build/ci_sanity.sh#L84

AndreasMadsen commented:
@drpngx I see, unfortunately it depends on a specific python version being in a specific location. I've committed the lint fixes, I hope they are correct but I have no way to be sure.

AndreasMadsen commented:
Fixed the spelling error too.

lukaszkaiser commented:
@tensorflow-jenkins test this please

lukaszkaiser commented:
One more thing, commas missing in the BUILD file. Hopefully that's the last thing. (And yes, we need a better process for that, I think.)

FAIL: buildifier found errors and/or warnings in above BUILD files.
buildifier suggested the following changes:
15c15
<     "tf_py_test"
---
>     "tf_py_test",
28a29
>         "//tensorflow/python:array_ops",
31d31
<         "//tensorflow/python:array_ops",

AndreasMadsen commented:
I found only one missing comma

lukaszkaiser commented:
@tensorflow-jenkins test this please

lukaszkaiser commented:
Please also move
  "//tensorflow/python:array_ops",
in the sparsemax/BUILD file 2 lines up (so that the order is alphabetic). Thanks!

AndreasMadsen commented:
Should be in alphabetical order now

lukaszkaiser commented:
@tensorflow-jenkins test this please

lukaszkaiser commented:
It looks like this part is ok now. The failing things look unrelated, I think there is a recent problem with testing. I'll re-test tomorrow, it might be all fine.

drpngx commented:
Thanks! We're having build issues that cause the system to timeout. We rebooted machines to see if the problem would go away.

lukaszkaiser commented:
@tensorflow-jenkins test this please

lukaszkaiser commented:
@tensorflow-jenkins test this please

drpngx commented:
We have new failures, including a test timeout.

```
14:14:18  11/189 Test  #28: C:/tf_jenkins/home/workspace/tensorflow-pr-win-cmake-py/tensorflow/python/kernel_tests/conv_ops_test.py ............................***Failed    6.81 sec
14:14:18 Traceback (most recent call last):
14:14:18   File "C:/tf_jenkins/home/workspace/tensorflow-pr-win-cmake-py/tensorflow/python/kernel_tests/conv_ops_test.py", line 32, in <module>
14:14:18     from tensorflow.contrib import layers
14:14:18   File "C:\Program Files\Anaconda3\lib\site-packages\tensorflow\contrib\__init__.py", line 52, in <module>
14:14:18     from tensorflow.contrib import sparsemax
14:14:18 ImportError: cannot import name 'sparsemax'
```

Can you `bazel test -c opt tensorflow/...`?

AndreasMadsen commented:
This is the output I get from running ` bazel test -c opt tensorflow/...` https://gist.github.com/AndreasMadsen/a11009cd3cda4d6058dfb4dee5072356

tl;dr it fails with:

```
ERROR: /private/var/tmp/_bazel_Andreas/99eb1091050df9ac5ca7bd7d8ea0247e/external/protobuf/BUILD:579:1: C++ compilation of rule '@protobuf//:internal/_api_implementation.so' failed: cc_wrapper.sh failed: error executing command external/local_config_cc/cc_wrapper.sh -U_FORTIFY_SOURCE '-D_FORTIFY_SOURCE=1' -fstack-protector -Wall -Wthread-safety -Wself-assign -fcolor-diagnostics -fno-omit-frame-pointer -g0 -O2 -DNDEBUG ... (remaining 41 argument(s) skipped): com.google.devtools.build.lib.shell.BadExitStatusException: Process exited with status 1.
external/protobuf/python/google/protobuf/internal/api_implementation.cc:31:10: fatal error: 'Python.h' file not found
#include <Python.h>
         ^
1 error generated.
```

I've just executed the tests individually with `python path_to_test.py`. I've also confirmed that `help(tf.contrib.sparsemax)` works.

drpngx commented:
Did you run bazel fetch?

On Jan 19, 2017 8:31 AM, "Andreas Madsen" <notifications@github.com> wrote:

> This is the output I get from running bazel test -c opt tensorflow/...
> https://gist.github.com/AndreasMadsen/a11009cd3cda4d6058dfb4dee5072356
>
> tl;dr it fails with:
>
> ERROR: /private/var/tmp/_bazel_Andreas/99eb1091050df9ac5ca7bd7d8ea0247e/external/protobuf/BUILD:579:1: C++ compilation of rule '@protobuf//:internal/_api_implementation.so' failed: cc_wrapper.sh failed: error executing command external/local_config_cc/cc_wrapper.sh -U_FORTIFY_SOURCE '-D_FORTIFY_SOURCE=1' -fstack-protector -Wall -Wthread-safety -Wself-assign -fcolor-diagnostics -fno-omit-frame-pointer -g0 -O2 -DNDEBUG ... (remaining 41 argument(s) skipped): com.google.devtools.build.lib.shell.BadExitStatusException: Process exited with status 1.
> external/protobuf/python/google/protobuf/internal/api_implementation.cc:31:10: fatal error: 'Python.h' file not found
> #include <Python.h>
>          ^
> 1 error generated.
>
> I've just executed the tests individually with python path_to_test.py.
> I've also confirmed that help(tf.contrib.sparsemax) works.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/6387#issuecomment-273824937>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AT_SbTQ2_4VNhOqrVvM8jkiZl6ZSlT4Hks5rT4_0gaJpZM4LQJo->
> .
>

AndreasMadsen commented:
No. I ran `bazel fetch tensorflow/...` now and gets the same error (though the `libtool` warnings no longer appears):

```
INFO: Found 1289 targets and 839 test targets...
ERROR: /private/var/tmp/_bazel_Andreas/99eb1091050df9ac5ca7bd7d8ea0247e/external/protobuf/BUILD:579:1: C++ compilation of rule '@protobuf//:internal/_api_implementation.so' failed: cc_wrapper.sh failed: error executing command external/local_config_cc/cc_wrapper.sh -U_FORTIFY_SOURCE '-D_FORTIFY_SOURCE=1' -fstack-protector -Wall -Wthread-safety -Wself-assign -fcolor-diagnostics -fno-omit-frame-pointer -g0 -O2 -DNDEBUG ... (remaining 41 argument(s) skipped): com.google.devtools.build.lib.shell.BadExitStatusException: Process exited with status 1.
external/protobuf/python/google/protobuf/internal/api_implementation.cc:31:10: fatal error: 'Python.h' file not found
#include <Python.h>
         ^
```

drpngx commented:
This is python development packages. On ubuntu, install the `python-dev` package.

AndreasMadsen commented:
@drpngx It is a `brew` install on MacOS and `python3-config --include` points to the correct directories.

drpngx commented:
Strange. How did you build before?

AndreasMadsen commented:
I just followed the [Setting up TensorFlow for Development](https://www.tensorflow.org/get_started/os_setup#setting_up_tensorflow_for_development) instructions.

```
bazel build -c opt //tensorflow/tools/pip_package:build_pip_package

mkdir _python_build
cd _python_build
ln -s ../bazel-bin/tensorflow/tools/pip_package/build_pip_package.runfiles/org_tensorflow/* .
ln -s ../tensorflow/tools/pip_package/* .
python3 setup.py develop
```

rmlarsen commented:
@tensorflow-jenkins test this please

rmlarsen commented:
@AndreasMadsen It looks like your test is timing out on CPU. Could you possibly split it up into smaller test cases that can be run in parallel to bring down the time?

AndreasMadsen commented:
I've reduced the number of test observations from `100` to `10`. I'm not sure what the criteria is for splitting up a test, can I simply split `testFloat` into `test_sparsemax_against_numpy_float`, `test_sparsemax_of_zero_float`, ...? I just followed the pattern used in other tests.

rmlarsen commented:
@AndreasMadsen I agree that it is probably difficult to split the tests more than they currently are. If the size reduction does the trick, I'm OK with merging this. Thanks for the quick response.

rmlarsen commented:
@tensorflow-jenkins test this please

rmlarsen commented:
@AndreasMadsen: The errors look unrelated, but I would like to wait and re-run the tests after the fixes in https://github.com/tensorflow/tensorflow/pull/7209 are merged to be safe. Thanks for your patience.

rmlarsen commented:
@tensorflow-jenkins test this please

rmlarsen commented:
@AndreasMadsen it looks like some of the errors are genuine, e.g.

10:19:22 Traceback (most recent call last):
10:19:22   File "C:/tf_jenkins/home/workspace/tensorflow-pr-win-cmake-py/tensorflow/python/kernel_tests/conv_ops_test.py", line 32, in <module>
10:19:22     from tensorflow.contrib import layers
10:19:22   File "C:\Program Files\Anaconda3\lib\site-packages\tensorflow\contrib\__init__.py", line 52, in <module>
10:19:22     from tensorflow.contrib import sparsemax
10:19:22 ImportError: cannot import name 'sparsemax'

I suspect you need to add an entry for sparsemax in the top-level tensorflow/BUILD file in the all_opensource_files filegroup.

AndreasMadsen commented:
@rmlarsen I see, thanks. I've added `sparsemax` to the `all_opensource_files` filegroup.

rmlarsen commented:
@tensorflow-jenkins test this please

rmlarsen commented:
@AndreasMadsen OK one more change (I hope): To make this work with our cmake build, you need to add your module here: 

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/cmake/tf_python.cmake#L375

Just follow the existing pattern.

AndreasMadsen commented:
Done

rmlarsen commented:
@tensorflow-jenkins test this please

rmlarsen commented:
@gunan AFAICT the remaining build error in unrelated to Andreas' code. Can you please confirm that it's OK to merge this?

gunan commented:
It looks like agent went down during build.
But as there are cmake changes in the PR, would it be ok to wait testing just windows cmake (We can merge when windows passes.
here is the build:
http://ci.tensorflow.org/job/tensorflow-pr-win-cmake-py/1060/

rmlarsen commented:
@gunan it looks like the build you kicked off finished successfully. I'll merge this now.

@AndreasMadsen thank you for the contribution. This is really nice.

AndreasMadsen commented:
Thanks for the continued guidance.
